### ISSUE TITLE
Update gradient descent for per-parameter learning rates

### ISSUE THREAD

**Issue by BradyPlanden** [pybop-team/PyBOP#567]
### Feature description

Add a per-parameter learning rate implementation of gradient descent.

### Motivation

The current gradient descent implementation uses the minimum `sigma0` applied to all parameters. This is problematic if the gradient magnitudes are dissimilar, or for learning rate calibration to fine tune convergence.
